repo/fsrepo/migrations/fetch.go

```diff
@@ -0,0 +1,272 @@
+package migrations
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"runtime"
+	"strings"
+)
+
+const (
+	// Current dirstibution to fetch migrations from
+	CurrentIpfsDist = "/ipfs/Qme8pJhBidEUXRdpcWLGR2fkG5kdwVnaMh3kabjfP8zz7Y"
+
+	envIpfsDistPath = "IPFS_DIST_PATH"
+
+	// Distribution
+	gatewayURL = "https://ipfs.io"
+	ipfsDist   = "/ipns/dist.ipfs.io"
+
+	// Maximum download size
+	fetchSizeLimit = 1024 * 1024 * 512
+)
+
+type limitReadCloser struct {
+	io.Reader
+	io.Closer
+}
+
+var ipfsDistPath string
+
+func init() {
+	SetIpfsDistPath("")
+}
+
+// SetIpfsDistPath sets the ipfs path to the distribution site.  If an empty
+// string is given, then the path is set using the IPFS_DIST_PATH environ
+// veriable, or the default dns link value if that is not defined.
+func SetIpfsDistPath(distPath string) {
+	if distPath != "" {
+		ipfsDistPath = distPath
+		return
+	}
+
+	if dist := os.Getenv(envIpfsDistPath); dist != "" {
+		ipfsDistPath = dist
+	} else {
+		ipfsDistPath = ipfsDist
+	}
+}
+
+// FetchBinary downloads an archive from the distribution site and unpacks it.
+//
+// The base name of the archive file, inside the distribution directory on
+// distribution site, may differ from the distribution name.  If it does, then
+// specify arcName.
+//
+// The base name of the binary inside the archive may differ from the base
+// archive name.  If it does, then specify binName.  For example, the following
+// is needed because the archive "go-ipfs_v0.7.0_linux-amd64.tar.gz" contains a
+// binary named "ipfs"
+//
+//     FetchBinary(ctx, "go-ipfs", "v0.7.0", "go-ipfs", "ipfs", tmpDir)
+//
+// If out is a directory, then the binary is written to that directory with the
+// same name it has inside the archive.  Otherwise, the binary file is written
+// to the file named by out.
+func FetchBinary(ctx context.Context, dist, ver, arcName, binName, out string) (string, error) {
```

Comment:
The current setup here is that we have a global function `FetchBinary` that internally calls `fetchHttp` and `fetchIpfs`, where `fetchIpfs` is using go-ipfs-api.

This is a bit unfortunate as it:
1. Adds a dependency on go-ipfs-api
2. Doesn't really leave us room for figuring out a way for go-ipfs to download the binary over IPFS without reworking things (i.e. when go-ipfs automatically does a migration it's not going to be able to use IPFS to handle this since the node isn't and can't be running yet)

One way we might clean this up is to have a `Fetcher` interface that just gets the archive and is implemented by an `HttpFetcher` and an `IpfsFetcher` and we can pass the Fetcher to the `FetchBinary` function. Additionally, we can have the `IpfsFetcher` take a CoreAPI and then callers like ipfs-update can use https://github.com/ipfs/go-ipfs-http-client as their CoreAPI implementation.

WDYT?

Reply:
I see your point regarding 1 and 2.  The Ipfs fetcher was to support `ipfs-update` and `fs-repo-migrations`, and I did not consider that for automated migrations.

As far as passing an `IpfsFetcher` or `HttpFetcher` to `FetchBinary`, that means that a decision has already been made to use one or the other.  With `ipfs-update`, I wanted to preserve its ability to try fetching from IPFS, and if not available, then fail-over to fetching from HTTP.  That could be solved by a "compountFetcher" implementation whose `Fetch` method tries one, then the other. 

The other issue with passing in a Fetcher is that I would also need to pass the Fetcher to every API that needs to fetch anything.  This includes fetching available and latest versions and running migrations.  Is it reasonable for all these APIs to require a Fetcher?

Reply:
> compoundFetcher

Yep, exactly my thinking.

>  Is it reasonable for all these APIs to require a Fetcher?

I think so, while we could use globals I'm generally a fan of explicitness even if it adds a bit of verbosity to the API.

Reply:
Fetcher is now passed into functions that need to fetch from network. ipfs-update will provide IpfsFetcher implementation of Fetcher interface.